### PR TITLE
178 feature/delete terms  by postfix prefix

### LINKED ISSUE #177: Delete Taxonomy Terms - Based on post count
- [x] Create a new module with the name 'Delete Taxonomy Terms by post count' in the Delete Terms page.
- [x] Choose the taxonomy. Single select dropdown. Create Helper functions in `Renderer`.
- [x] Choose the post type. Single select dropdown.
- [x] A dropdown to choose the operation. Possible values (not equal to, equal to, less than, greater than)
- [x] A textbox to enter the post count.

### LINKED ISSUE #178: Delete Taxonomy Terms - Based on postfix and prefix
- [x] Create a new Delete Terms page.
- [x] Ability to choose a taxonomy
- [x] Ability to choose terms. Use the following format Delete terms that |select| with |textbox| (equal to, not equal to, starts with, ends with, contains, non contains)
- [ ] Filter: Only if it doesn't contain any post
- [x] Change name to 'Delete Taxonomy Terms by Name'

### LINKED ISSUE #348: Delete Taxonomy terms by prefix and postfix - Add tests

### LINKED ISSUE #349: Delete Taxonomy terms by post count - Add tests

### PR DESCRIPTION
**Pull request by rajanit2000**
Has code for the following issues.

Fix #177 
Fix #348 
Fix #178 
Fix #349 

## Delete Terms based on Post Count

- [x] Fix all test case issues for Delete Terms based on Post count. Add additional terms that should not be deleted.
- [x] Add tests to check deletion of terms from custom taxonomy

## Delete Terms based on Name

- [x] Add test cases to verify deletion of terms based on prefix or postfix.
- [x] Add tests to check deletion of terms from custom taxonomy

### CODE REVIEW

**Comment by rajanit2000** [comment]
@sudar this PR have two issues #177 and #178 
Its ready to review

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+ * @since 6.0.0
+ */
+class DeleteTermsByPostCountModule extends TermsModule {
+	protected function initialize() {
+		$this->item_type     = 'terms';
+		$this->field_slug    = 'by_post_count';
+		$this->meta_box_slug = 'bd_by_post_count';
+		$this->action        = 'delete_terms_by_post_count';
+		$this->cron_hook     = 'do-bulk-delete-term-post-count';
+		$this->scheduler_url = 'http://bulkwp.com/addons/scheduler-for-deleting-posts-by-category/?utm_source=wpadmin&utm_campaign=BulkDelete&utm_medium=buynow&utm_content=bd-sc';
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: This url is wrong. Change it to an empty string.

File: include/Core/Terms/DeleteTermsPage.php [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	 *
+	 * @param array $help_tabs Help tabs.
+	 *
+	 * @return array Modified list of tabs.
+	 */
+	protected function add_help_tab( $help_tabs ) {
+		$overview_tab = array(
+			'title'    => __( 'Overview', 'bulk-delete' ),
+			'id'       => 'overview_tab',
+			'content'  => '<p>' . __( 'This screen contains different modules that allows you to delete posts or schedule them for deletion.', 'bulk-delete' ) . '</p>',
```
Comment (sudar) [inline include/Core/Terms/DeleteTermsPage.php]: This description is wrong.

Change it to 

> This screen contains different modules that allows you to delete terms from taxonomies

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * @since 6.0.0
+ */
+class DeleteTermsByPostCountModule extends TermsModule {
+	protected function initialize() {
+		$this->item_type     = 'terms';
+		$this->field_slug    = 'by_post_count';
+		$this->meta_box_slug = 'bd_by_post_count';
+		$this->action        = 'delete_terms_by_post_count';
+		$this->cron_hook     = 'do-bulk-delete-term-post-count';
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `do-bulk-delete-term-by-post-count`

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+/**
+ * Delete Terms by Post Count.
+ *
+ * @since 6.0.0
+ */
+class DeleteTermsByPostCountModule extends TermsModule {
+	protected function initialize() {
+		$this->item_type     = 'terms';
+		$this->field_slug    = 'by_post_count';
+		$this->meta_box_slug = 'bd_by_post_count';
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `bd_delete_terms_by_post_count`

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+/**
+ * Delete Terms by Post Count.
+ *
+ * @since 6.0.0
+ */
+class DeleteTermsByPostCountModule extends TermsModule {
+	protected function initialize() {
+		$this->item_type     = 'terms';
+		$this->field_slug    = 'by_post_count';
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `terms_by_post_count`

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+		$this->meta_box_slug = 'bd_by_post_count';
+		$this->action        = 'delete_terms_by_post_count';
+		$this->cron_hook     = 'do-bulk-delete-term-post-count';
+		$this->scheduler_url = 'http://bulkwp.com/addons/scheduler-for-deleting-posts-by-category/?utm_source=wpadmin&utm_campaign=BulkDelete&utm_medium=buynow&utm_content=bd-sc';
+		$this->messages      = array(
+			'box_label'  => __( 'By Terms by Post Count', 'bulk-delete' ),
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `Delete Terms by Post Count`

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+			'scheduled'  => __( 'The selected posts are scheduled for deletion', 'bulk-delete' ),
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `The selected terms are scheduled for deletion`

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+		);
+	}
+
+	/**
+	 * Render Delete terms by postfix and prefix box.
+	 */
+	public function render() {
+		?>
+		<!-- Category Start-->
+		<h4><?php _e( 'Select the taxonomy from which you want to delete', 'bulk-delete' ); ?></h4>
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change this to `Select the taxonomy from which you want to delete terms`

File: include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Delete Terms by Postfix and Prefix.
+ *
+ * @since 6.0.0
+ */
+class DeleteTermsByPostfixAndPrefixModule extends TermsModule {
+	protected function initialize() {
+		$this->item_type     = 'terms';
+		$this->field_slug    = 'by_name';
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php]: Do all the changes that I have specified in the `DeleteTermsByPostCountModule`

File: include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+				<?php $this->render_taxonomy_dropdown(); ?>
+			</table>
+
+			<table class="optiontable">
+				<?php $this->render_term_options(); ?>
+			</table>
+
+			<table class="optiontable">
+				<?php
+				// $this->render_have_post_settings(); // TODO
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php]: Remove commented code.

File: include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+				$query['name__like'] = $term_text;
+				break;
+
+			case 'not_equal_to':
+				$term_ids         = bd_term_query( array( 'name__like' => $term_text ), $options['taxonomy'] );
+				$query['exclude'] = $term_ids;
+				break;
+
+			case 'starts':
+				$term_ids            = $this->bd_term_starts( $term_text , $options );
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php]: Remove `bd_` prefix from all the methods.

File: include/Core/Terms/TermsModule.php [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+			}else{
+				wp_delete_term( $term_id, $options['taxonomy'] );
+				$count++;
+			}
+		}
+
+		return $count;
+	}
+
+	protected function bd_starts_with($haystack, $needle){
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Write phpdoc for all the helper functions.

File: include/util/query.php [hunk 13]
```diff
@@ -102,3 +102,51 @@ function bd_query( $options ) {
...
+ * Adds some performance enhancing defaults.
+ *
+ * @since  6.0
+ *
+ * @param array $options  List of options
+ * @param mixed $taxonomy
+ *
+ * @return array Result array
+ */
+function bd_term_query( $options, $taxonomy ) {
```
Comment (sudar) [inline include/util/query.php]: Move this function to `TermsModule` class.

**Comment by sudar** [comment]
@rajanit2000 

I have left a couple of review comments for this branch.

I have also merged latest `dev/6.0.0` into this branch, so you may have to pull in the latest code first.

**Comment by rajanit2000** [comment]
@sudar TravisCI shows some npm ERR. Can you please guide me how to fix this issue?
https://travis-ci.org/sudar/bulk-delete/jobs/397559337

**Comment by sudar** [comment]
@rajanit2000 

I have merged the latest `dev/6.0.0` branch into your branch and this should solve the Travis issue that you were facing.

Let me know if this still affects you.

**Comment by rajanit2000** [comment]
@sudar help me to how can i fix this issue?
https://app.codacy.com/app/sudar/bulk-delete/pullRequest?prid=1846922

**Comment by rajanit2000** [comment]
Status: 90% done, Need to fix some codacy changes

File: include/Core/Terms/TermsModule.php [hunk 14]
```diff
@@ -0,0 +1,294 @@
...
+	/**
+	 * Build the query using query params and then Delete posts.
+	 *
+	 * @param array $query   Params for WP Query.
+	 * @param array $options Delete Options.
+	 *
+	 * @return int Number of posts deleted.
+	 */
+	protected function delete_terms_from_query( $query, $options ) {
+		$term_ids = bd_term_query( $query, $options['taxonomy'] );
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: The function bd_term_query was not found.

File: include/Core/Terms/TermsModule.php [hunk 15]
```diff
@@ -0,0 +1,294 @@
...
+	 *
+	 * @return int Number of posts deleted.
+	 */
+	protected function delete_terms_by_id( $term_ids, $options ) {
+		$count = 0;
+
+		foreach ( $term_ids as $term_id ) {
+			$term = get_term( $term_id, $options['taxonomy'] );
+
+			if ( isset( $options['no_posts'] ) && $term->count > 0 ) {
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Make sure `$term` is not `WP_Error`

File: include/Core/Terms/TermsModule.php [hunk 16]
```diff
@@ -0,0 +1,294 @@
...
+		return $count;
+	}
+
+	/**
+	 * custom string function use to get is string start with specified string.
+	 *
+	 * @param string $haystack.
+	 * @param string $needle.
+	 *
+	 * @return boolean.
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Change data type to `bool`.

**Comment by rajanit2000** [comment]
@sudar its finished, ready to review

File: include/Core/Terms/TermsModule.php [hunk 17]
```diff
@@ -0,0 +1,302 @@
...
+	protected function delete_terms_from_query( $query, $options ) {
+		$term_ids = $this->term_query( $query, $options['taxonomy'] );
+
+		return $this->delete_terms_by_id( $term_ids, $options );
+	}
+
+	/**
+	 * Render the "private post" setting fields.
+	 */
+	protected function render_private_post_settings() {
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Move this function to `Renderer` class.

File: include/Core/Terms/TermsModule.php [hunk 18]
```diff
@@ -0,0 +1,302 @@
...
+			$posts = get_posts($args);
+
+			$term_ids[] = $this->get_term_id_by_name( $options['term_text'], $options['term_opt'], $term->term_id, count($posts) );
+		}
+
+		return $term_ids;
+	}
+
+	protected function get_term_id_by_name( $term_text, $term_opt, $term_id, $post_count ){
+		if( $post_count == $term_text && $term_opt == 'equal_to' ){
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: In the call, I asked you to use switch case instead of if/else.

**Comment by rajanit2000** [comment]
Codacy still is not passed, But want to review 
Codecy issue something we want to ignore it 
https://app.codacy.com/app/rajanit2000/bulk-delete/pullRequest?prid=1875796

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

I couldn't understand what the following module does with the description. Should we change the descriptions may be?

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

The `With` keyword is confusing. I think you should remove it.

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

Fix the spelling.

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

What will happen to posts if the terms are deleted? I think it is better to add a description about that.

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

Should one enter Term name or Term slug? Kindly add a description/label to make users input what you expect.

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

Kindly fix the message in the alert box. We aren't deleting posts.

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

Let's be consistent with our warnings.

image

image

**Comment by mariadanieldeepak** [comment]
@rajanit2000 - Another review comment posted [here](https://github.com/sudar/bulk-delete/issues/178#issuecomment-407279603)

**Comment by rajanit2000** [comment]
@mariadanieldeepak is it possible please post the suggested content for each comment

**Comment by rajanit2000** [comment]
@mariadanieldeepak I have done the UI changes, And i can't find the issue 

https://github.com/sudar/bulk-delete/issues/178#issuecomment-408635923

If am wrong please explain the issue

**Comment by mariadanieldeepak** [comment]
@rajanit2000,

Could you merge the latest `dev/6.0.0` in to your branch?

File: include/Core/Terms/TermsModule.php [hunk 19]
```diff
@@ -0,0 +1,307 @@
+<?php
```
Comment (mariadanieldeepak) [inline include/Core/Terms/TermsModule.php]: @rajanit2000, Fix PHPCS issues on this file. Currently there are 148 errors.

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php [hunk 20]
```diff
@@ -0,0 +1,202 @@
+<?php
```
Comment (mariadanieldeepak) [inline tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php]: @rajanit2000, 

There are 15 Code sniffer errors on this file too. Could you get it fixed?

File: include/Core/Base/Mixin/Renderer.php [hunk 21]
```diff
@@ -111,4 +111,61 @@ protected function render_user_role_dropdown() {
 	protected function render_post_type_dropdown() {
 		bd_render_post_type_dropdown( $this->field_slug );
 	}
+
```
Comment (mariadanieldeepak) [inline include/Core/Base/Mixin/Renderer.php]: @rajanit2000, 

Also, verify if you're sticking to standards in this file too.

**Comment by rajanit2000** [comment]
@mariadanieldeepak I have update the phpcs issues, Now its ready to review

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 22]
```diff
@@ -0,0 +1,125 @@
...
+			'cron_label' => __( 'Delete Terms By Post Count', 'bulk-delete' ),
+		);
+	}
+
+	/**
+	 * Render Delete terms by postfix and prefix box.
+	 */
+	public function render() {
+		?>
+		<!-- Category Start-->
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Wrong comment. Change it.

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+			</table>
+
+			<table class="optiontable">
+				<?php $this->render_term_options(); ?>
+			</table>
+
+		</fieldset>
+		<?php
+		$this->render_submit_button();
+	}
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Add a blank line after the method.

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php [hunk 24]
```diff
@@ -0,0 +1,125 @@
...
+	/**
+	 * Response message for deleting posts.
+	 *
+	 * @param int $items_deleted Total number of posts deleted.
+	 *
+	 * @return string Response message
+	 */
+	protected function get_success_message( $items_deleted ) {
+		/* translators: 1 Number of posts deleted */
+		return _n( 'Deleted %d term with the selected options', 'Deleted %d terms with the selected terms count', $items_deleted, 'bulk-delete' );
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Change the success message.

File: include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+	/**
+	 * Response message for deleting posts.
+	 *
+	 * @param int $items_deleted Total number of posts deleted.
+	 *
+	 * @return string Response message
+	 */
+	protected function get_success_message( $items_deleted ) {
+		/* translators: 1 Number of posts deleted */
+		return _n( 'Deleted %d term with the selected options', 'Deleted %d terms with the selected terms', $items_deleted, 'bulk-delete' );
```
Comment (sudar) [inline include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php]: Change the success message.

File: include/Core/Terms/TermsModule.php [hunk 26]
```diff
@@ -0,0 +1,359 @@
...
+	 *
+	 * @return int Number of posts deleted.
+	 */
+	protected function delete_terms_by_id( $term_ids, $options ) {
+		$count = 0;
+
+		foreach ( $term_ids as $term_id ) {
+			$term = get_term( $term_id, $options['taxonomy'] );
+
+			if ( is_wp_error( $term ) ) {
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Change this check to

`if ( ! $term instanceof \WP_Term )`

File: include/Core/Terms/TermsModule.php [hunk 27]
```diff
@@ -0,0 +1,359 @@
...
+		return $count;
+	}
+
+	/**
+	 * Custom string function use to get is string start with specified string.
+	 *
+	 * @param string $haystack search string.
+	 * @param string $needle   find string.
+	 *
+	 * @return bool.
```
Comment (sudar) [inline include/Core/Terms/TermsModule.php]: Change this to boolean.

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php (hunk 22 above)
Comment (rajanit2000) [reply include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Fixed

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php (hunk 23 above)
Comment (rajanit2000) [reply include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Fixed

File: include/Core/Terms/Modules/DeleteTermsByPostCountModule.php (hunk 24 above)
Comment (rajanit2000) [reply include/Core/Terms/Modules/DeleteTermsByPostCountModule.php]: Fixed

File: include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php (hunk 25 above)
Comment (rajanit2000) [reply include/Core/Terms/Modules/DeleteTermsByPostfixAndPrefixModule.php]: fixed

File: include/Core/Terms/TermsModule.php (hunk 26 above)
Comment (rajanit2000) [reply include/Core/Terms/TermsModule.php]: Fixed

File: include/Core/Terms/TermsModule.php (hunk 27 above)
Comment (rajanit2000) [reply include/Core/Terms/TermsModule.php]: Fixed

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByNameModuleTest.php [hunk 28]
```diff
@@ -0,0 +1,257 @@
...
+	public function setUp() {
+		parent::setUp();
+
+		$this->module = new DeleteTermsByNameModule();
+	}
+
+	/**
+	 * Dataprovider of test case.
+	 */
+	public function delete_terms_by_name_various_inputs() {
```
[the remark this answers is not in the document]
Comment (sudar) [reply tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByNameModuleTest.php]: The name should be of the format `provide_data_to_<function_name>`

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByNameModuleTest.php [hunk 29]
```diff
@@ -0,0 +1,257 @@
...
+
+		// call our method.
+		$delete_options = array(
+			'taxonomy'  => $taxonomy,
+			'term_opt'  => $operation['operator'],
+			'term_text' => $operation['term_text'],
+		);
+		$deleted_term   = $this->module->delete( $delete_options );
+
+		$this->assertEquals( $expected, $deleted_term );
```
Comment (sudar) [inline tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByNameModuleTest.php]: Explicitly query the term that will not be deleted and assert that.

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php [hunk 30]
```diff
@@ -0,0 +1,454 @@
...
+	public function setUp() {
+		parent::setUp();
+
+		$this->module = new DeleteTermsByPostCountModule();
+	}
+
+	/**
+	 * Dataprovider of test case.
+	 */
+	public function delete_terms_by_post_count_various_inputs() {
```
[the remark this answers is not in the document]
Comment (sudar) [reply tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php]: Change the method name to follow the format.

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php [hunk 31]
```diff
@@ -0,0 +1,454 @@
...
+			),
+
+			array(
+				array(
+					'post_type'  => 'post',
+					'taxonomy'   => 'post_tag',
+					'term_input' => array(
+						array(
+							'term'            => 'Term A',
+							'number_0f_posts' => 2,
```
Comment (sudar) [inline tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php]: Check if it is zero or `o`

File: tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php [hunk 32]
```diff
@@ -0,0 +1,454 @@
...
+			'post_type' => array( $post_type ),
+			'taxonomy'  => $taxonomy,
+			'term_opt'  => $operation['operator'],
+			'term_text' => $operation['post_count'],
+		);
+		$deleted_term   = $this->module->delete( $delete_options );
+
+		$this->assertEquals( $expected, $deleted_term );
+
+	}
```
Comment (sudar) [inline tests/wp-unit/include/Core/Terms/Modules/DeleteTermsByPostCountModuleTest.php]: Fetch the term that will not be deleted and assert that.

**Comment by rajanit2000** [comment]
@sudar Its ready to review.

**Pull request merged**